### PR TITLE
0.5.0 dev

### LINKED ISSUE #240: Bug with package name matching
See https://github.com/ember-cli-deploy/ember-cli-deploy-lightning-pack/issues/6#issuecomment-148884932

### PR DESCRIPTION
**Pull request by lukemelia**
This PR is for the implementation of the deploy pipeline which will make up `v0.5.0`. It was previously in #130.

Overview of the new pipeline plus plugins architecture for 0.5.0: https://github.com/ember-cli/ember-cli-deploy/blob/0.5.0-dev/rfc/plugins.md

Note: for people reviewing this, you'll also want to see the new plugins linked in the tasks below.

If you would like to help, and are not already part of the ember-cli-deploy slack channel, open an issue to get an invite.

TODO:
- [x] Tighten up console output logging conventions (eg, colours, standard output types, verbosity) (@lukemelia)
- [x] fix #240 (@lukemelia)
- [x] fill in `Getting Started > Quickstart` section in doco site. (@achambers)
- [x] update usage documentation (README) and usage info on website for ember-cli-deploy
- [x] Make `ember-deploy-s3-index` 0.5.0 plugin compatible (@LevelbossMike)
- [x] expand `Plugins > Writing a plugin` with extra configuration (@ghedamat)
- [x] Upgrade documentation (@samselikoff/@duizendnegen #189)
- [x] document new slack plugin; should follow standard plugin format 
- [x] fill in `Getting Started > Overview` section in doco site. (@achambers)
- [x] fill in `Getting Started > Installation` section in doco site. (@achambers)
- [x] fill in `The Deploy Pipeline > Overview` section in doco site. @ghedamat
- [x] fill in `The Deploy Pipeline > Pipeline Hooks` section in doco site. (@ghedamat)
- [x] fill in `The Deploy Pipeline > The Deployment Context` section in doco site. (@ghedamat)
- [x] fill in `Configuration > Overview` section in doco site. (@ghedamat)
- [x] fill in `Configuration > Aliasing Plugins` section in doco site. (@ghedamat)
(https://github.com/LevelbossMike/ember-cli-deploy-slack) (@LevelbossMike)
- [x] Post install message that says something along the lines of "ember-cli-deploy needs plugins to actually do the deployment work. Seen [URL] to learn how to install plugins and see what plugins are available." (@ghedamat)
- [x] fill in `Plugins > Overview` section in doco site. @ghedamat
- [x] fill in `Plugins > Plugins` section in doco site.
- [x] tidy up `Plugins > Plugins` section in doco site.
- [x] fill in `Plugins > Plugin packs` section in doco site.
- [x] fill in `Plugins > Writing a plugin` section in doco site.
- [x] fill in `Usage > Overview` section in doco site.  (@ghedamat)
- [x] Document the `didFail` hook (@ghedamat)
- [x] fill in `Configuration > dotEnv support` section in doco site.
- [x] document dotEnv support (@achambers)
- [x] create documentation for plugin authors (@achambers)
- [x] document plugin packs (@achambers)
- [x] website section to list deploy plugins
- [x] emit warning when pipeline executed without any plugins found (@achambers)
- [x] Rebase this branch on master
- [x] work with 0.4.x plugin authors to ensure they are ready for 0.5.0 
- [x] Enable 0.5.x docs site section
- [x] figure out if we should move plugins to ember-cli or new ember-cli-deploy org
- [x] Implement [this](https://github.com/zapnito/ember-cli-deploy-revision-key/issues/14) (@achambers)
- [x] Add new `prepare` hooks between `build` and `upload` (@achambers)
- [x] Remove code in `included` hook that uses 'fingerprint' options (@achambers)
- [x] Remove broccoli-asset-rev dependency in revision-key plugin (@achambers)
- [x] Cleanup ember-cli-deploy project, removing no longer used files (@lukemelia #197)
- [x] release beta to npm
- [x] document new gzip plugin (https://github.com/lukemelia/ember-cli-deploy-gzip) (@achambers)
- [x] release ecd-s3 to npm
- [x] release ecd-redis to npm
- [x] release ecd-revision-key to npm
- [x] release ecd-json-config to npm
- [x] release ecd-build to npm
- [x] Pipeline implementation (@levelbossmike & @achambers)
- [x] implement gzip plugin that works nicely with s3 plugin (@lukemelia)
- [x] implement manifest plugin and support in s3 plugin for differential deployment (@lukemelia)
- [x] implement slack notification plugin (@levelbossmike)
- [x] Add a **didFail** hook for plugins to undo their changes on a broken deployment (@levelbossmike)
- [x] Add `configure` hook to pipeline and refactor plugins (@achambers)
- [x] Implement `activate` command as a pipeline (@achambers)
- [x] Implement `activate` hook in `ember-cli-deploy-redis` (@achambers)
- [x] Deploy command's pipeline should only include activate hooks if activate is enabled via CLI flag or configuration (@achambers)
- [x] Refactor pipeline task so it only registers the plugin hooks for the hooks the pipeline was initialised with (@achambers)
- [x] Implement "plugin pack" support and "lightning" pack (@lukemelia)
- [x] Implement hook in redis plugin to allow logging of preview URL when deploying a revision without activating. (@lukemelia)
- [x] common base class and or library for plugins (@lukemelia)
- [x] document new manifest plugin (https://github.com/lukemelia/ember-cli-deploy-manifest) (@lukemelia)
- [x] Give plugin authors a cleaner way to output to console that stays in line with the output formatting
- [x] Fix `deploy-config` blueprint so that it generates it in the format that works for 0.5.0 (ie, export a function instead of an object)
- [x] Implement specifying deploy environment to commands as a bare arg (e.g. `ember deploy qa`, `ember deploy:activate qa --revision=abc123etc`)
- [x] implement support for giving plugins alternate names via `plugins` config property (@dschmidt #194)
- [x] Implement list command as a pipeline (@duizendnegen #192)
- [x] document new redis plugin (https://github.com/zapnito/ember-cli-deploy-redis) (@achambers)
- [x] update redis plugin to support database index config option (@dschmidt)
- [x] document new revision-key plugin (https://github.com/zapnito/ember-cli-deploy-revision-key) (@achambers)
- [x] document new json-config plugin (https://github.com/zapnito/ember-cli-deploy-json-config) (@achambers)
- [x] document new lightning plugin pack (https://github.com/lukemelia/ember-cli-deploy-lightning-pack)
- [x] document new build plugin (https://github.com/zapnito/ember-cli-deploy-build) (@achambers)
- [x] document new s3 plugin (https://github.com/zapnito/ember-cli-deploy-s3) (@achambers)

### CODE REVIEW

**Comment by duizendnegen** [comment]
great to see progress towards 0.5.0 - I'd be interested in helping out
would be cool to, like for 0.4.0, have a "work with addon authors" step, so they can prepare, especially if you guys are going to make any endorsements ;)

**Comment by lukemelia** [comment]
@duizendnegen good idea. I added a task to the list. There is much to do, and we would love your help. Can you contact me at luke __at__ lukemelia.com with an email I can use to invite you to the ember-cli-deploy implementer's Slack team?

**Comment by achambers** [comment]
> Make ember-deploy-s3-index 0.5.0 plugin compatible (@LevelbossMike)

@LevelbossMike is the above task done? I know you've built something. Noticed there are zero tests. Just wondering if it's something we can tick off this list or whether there is more to do.

**Comment by LevelbossMike** [comment]
@achambers I still need to add some tests for that (which I am going to do this friday/weekend) after that I'll tick it off

File: UPGRADE_TO_0.5.0.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+1. update package.json keywords to include the keyword "ember-cli-deploy-plugin"
+2. `npm install ember-cli-deploy-plugin --save` and subclass the new https://github.com/lukemelia/ember-cli-deploy-plugin for all your adapters.
+3. implement `createDeployPlugin` in index.js and return an extended `ember-cli-deploy-plugin` from there.
+4. listen to hooks to actually upload stuff, most notably upload (index and assets adapter) and activate (index adapter)
+5. instruct your users to update their config (see above)
+6. instruct your users to install ember-cli-deploy-build (see above) (there's discussion of whether to put this in ember-cli-deploy by default for 0.5.0, follow https://github.com/ember-cli/ember-cli-deploy/pull/179)
+7. replace console.log statements to this.log to play nicely with the formatting in ember-cli-deploy
+
+### For index/store adapters:
+1. instruct your users to install ember-cli-deploy-revision-key (see above)
```
Comment (lukemelia) [inline UPGRADE_TO_0.5.0.md]: should be revision-data

**Comment by lukemelia** [comment]
:rocket: :rocket: :rocket: Amazing work on this everyone. Master is now the upcoming 0.5.0 release!

**Comment by dschmidt** [comment]
Woah :beers: :metal: :beers:

**Comment by ghedamat** [comment]
aww yeah

**Pull request merged**